lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,573 @@
+                                CP_DRY_AIR)
+
+
+class ExtractData(object):
+    """
+    A series of methods for extracting data from grid points to derive
+    diagnostic values at off grid positions.
+
+    """
+
+    def __init__(self, method='use_nearest'):
+        """
+        The class is called with the desired method to be used in extracting/
+        interpolating data to the site of interest from gridded data. If no
+        method is defined in the call to the class it defaults to the
+        'use_nearest' method.
+
+        """
+        self.method = method
+
+    def process(self, cube, sites, neighbours, ancillary_data, additional_data,
+                no_neighbours=9, lower_level=1, upper_level=3):
+        """
+        Call the correct function to enact the method of data extraction
+        specified. This function also handles multiple timesteps, consolidating
+        the resulting data cubes into an Iris.CubeList.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            Cube of diagnostic data from which to extract spotdata.
+
+        sites : dict
+            A dictionary containing the properties of spotdata sites.
+
+        neighbours : numpy.array
+            Array of neighbouring grid points that are associated with sites
+            in the SortedDictionary of sites.
+
+        additional_data : dict
+            A dictionary containing any supplmentary time varying diagnostics
+            that are needed for the selected extraction method.
+
+        ancillary_data : dict
+            A dictionary containing additional model data that is needed.
+            e.g. {'orography': <cube of orography>}
+
+        no_neighbours : int
+            Number of grid points about the site to consider when calculating a
+            local temperature lapse rate using orography variation.
+            e.g. consider a 5x5 grid of points -> no_neighbours = 25.
+
+        Returns:
+        --------
+        cube : iris.cube.Cube
+            An irregularly (i.e. non-gridded) cube of diagnostic data extracted
+            at the spotdata sites.
+
+        """
+        if self.method == 'use_nearest':
+            return self.use_nearest(cube, sites, neighbours)
+        elif self.method == 'orography_derived_temperature_lapse_rate':
+            orography = data_from_dictionary(ancillary_data, 'orography').data
+            return self.orography_derived_temperature_lapse_rate(
+                cube, sites, neighbours, orography,
+                no_neighbours=no_neighbours)
+        elif self.method == 'model_level_temperature_lapse_rate':
+
+            pressure_on_height_levels = (
+                data_from_dictionary(
+                    additional_data, 'pressure_on_height_levels')
+                )
+            # Ensure that pressure units are in Pa.
+            pressure_on_height_levels.convert_units('Pa')
+
+            lower_pressure = pressure_on_height_levels[lower_level, ...]
```

Comment:
It is making an assumption that the user knows what levels they want. As the levels can be set at runtime it is a decision that should be made with knowledge of the level set.